sound/soc/sof/ipc4-topology.c

```diff
@@ -1462,6 +1463,16 @@ static int sof_ipc4_widget_setup(struct snd_sof_dev *sdev, struct snd_sof_widget
 
 		msg = &pipeline->msg;
 		msg->primary |= pipeline->mem_usage;
+
```

Comment:
how about a pcm stream composed of many pipelines ? there will be only one scheduler so only need to allocate one ? how about others ?

Reply:
> how about a pcm stream composed of many pipelines ? there will be only one scheduler so only need to allocate one ? how about others ?

`sof_ipc4_widget_setup` will be triggered for each pipeline not just one scheduler. For example, we have two pipelines for SSP playback one for FE and the other for BE, `sof_ipc4_widget_setup` will be triggered for both pipelines and different pipeline (instance) id will be assigned.

Reply:
but the allocation code is under "case snd_soc_dapm_scheduler:" ?

Reply:
@RanderWang widgets in a pipeline are setup only after the pipeline widget for the pipeline they belong to are set up. So this code is fine.